HW2/P2/parallel_vector.pyx

```diff
@@ -70,23 +70,36 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
                              np.int32_t[:] src,
                              np.int32_t[:] dest,
                              int repeat):
-   cdef:
-       int idx, r
-       omp_lock_t *locks = get_N_locks(counts.shape[0])
-
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for each element of counts to parallelize
-   # data movement.  Be sure to avoid deadlock, and double-locking.
-   ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(counts.shape[0], locks)
+    cdef:
+        int idx, r
+        omp_lock_t *locks = get_N_locks(counts.shape[0])
+        np.int32_t min_idx, max_idx
+
+    ##########
+    # Your code here
+    # Use parallel.prange() and a lock for each element of counts to parallelize
+    # data movement.  Be sure to avoid deadlock, and double-locking.
+    ##########
+    with nogil:
+        for r in range(repeat):
+            for idx in prange(src.shape[0], num_threads=4):
+                min_idx = min(src[idx], dest[idx])
+                max_idx = max(src[idx], dest[idx])
+                
+                if min_idx == max_idx:
+                    continue
+
```

Comment:
Agreed. I got very confused here as there is not many comments.. so the problem asks us to acquire locks for each dest/src, I am not sure if you did right here. But what I did was as follow:
*********************************
acquire(&(locks[dest[idx]]));
+               counts[dest[idx]] += 1
+               release(&(locks[dest[idx]]));
+               acquire(&(locks[src[idx]]));
+               counts[src[idx]] -= 1
+               #free the lock here
+               release(&(locks[src[idx]]));
**************************************